### DISCUSSION TITLE
Reuse Rodauth config in custom Mailer

### DISCUSSION THREAD

**Discussion by benoror** [janko/rodauth-rails#143] [Ideas]
Follow up to: https://github.com/jeremyevans/rodauth/issues/251

To avoid that divergence, wondering if it would be good idea to keep default Mailers behaviors as close as possible to Rodauth's, for example:

`to:` field and `body:` via a proxy instance var:

```diff
-    @email_link = email_link(name, :verify_account, account_id, key)
+    @email_link = rodauth(name).verify_account_email_link
+    @body = rodauth(name).verify_account_email_body
+
-    mail to: @account.email, subject: rodauth(name).verify_account_email_subject
+    mail to: rodauth(name).email_to, subject: rodauth(name).verify_account_email_subject
```
https://github.com/janko/rodauth-rails/blob/main/lib/generators/rodauth/templates/app/mailers/rodauth_mailer.rb#L6

```diff
- Someone has created an account with this email address.  If you did not create
- this account, please ignore this message.  If you created this account, please go to
- <%= @email_link %>
+ <%= @body %>
- to verify the account.
```
https://github.com/janko/rodauth-rails/blob/main/lib/generators/rodauth/templates/app/views/rodauth_mailer/verify_account.text.erb

**Comment by janko** [comment]
Initially, `rodauth:install` didn't generate the mailer, so all email headers were set from Rodauth config, and you could run a separate `rodauth:mailer` generator. However, I had decided to [merge it into `rodauth:install`](https://github.com/janko/rodauth-rails/commit/789051595738e48e033451f70acb740f9df65481), because when someone wants to change the default body, it's much easier to do so via mailer templates.

Initially, the mailer received all parameters needed for the email body in arguments, but that meant potentially exposing secrets such as tokens and emaila via logs, and it didn't work with API + SPA that live on different domains. So, the mailer arguments now only receive raw tokens and no email addresses, so the mailer needs to do more work.

Note that the code changes to do what you want would have to be more complicated than what you've shown, as certain instance variables need to be set. I found it to require less code to reimplement part of the logic, and I'm relying on developers looking at the generated mailer, which should make it clear to them that Rodauth config such as `*_body` or `email_to` won't apply.

Since this is not a bug, I'm moving it to discussions.

**Reply by benoror** [reply]
now after using it further to customize emails actually I think it that makes more sense to generate a boilerplate for mailers instead

**Comment by janko** [comment]
After giving it some more thought, I'm considering going back to generating the mailer config separate from `rodauth:install`. I agree it makes sense that by default Rodauth config for emails is respected, and diverging from it would be a conscious decision. The new `rodauth:mailer` generator would print out config to copy-paste into the Rodauth configuration.

I also managed to rewrite the generated mailer to reuse `email_to`, and only replace `*_body` calls with templates. It ended up being cleaner to stay within Rodauth land, and I prefer it as it sets an example for other web frameworks. I will push that change shortly.

I'm not sure how much extracting the mailer out of `rodauth:install` would be considered a backwards incompatible change.

**Reply by benoror** [reply]
having separate explicit generators sounds like a sane idea in paper, but I am less concerned now that I understand the reasoning behind generating the mailers by default in the first place
